Fixes for externally created and deleted files.

TowelSniffer opened the pull request:
I'm not very experienced so I apologize if these suggestions don't make sense but they seem to work with my own testing. 

1. Build fix - Downgraded chokidar to ^3.6.0 in package.json:
chokidar v4.0 was causing build failures with tiddlywiki-plugin-dev

2. Duplicate _1 files when externally creating a .tid file: 
patched $tw.boot.files before addTiddler to prevent race conditions on file creation.

3. external deletions not updating in browser: 
re-added deleteTiddler so the browser ui can update with external deletions, and include check to handle server crashes.

4. Crashes with deleteTiddler: 
added safety check to prevent the crashing on unknown file deletions error that you mention. This seems to work for me but i could be missing something. 

5. Zombie draft.of tiddlers: 
added logic to ignore draft.of tiddlers. I was getting issues with zombie drafts remaining after creating new tiddlers, and it seems to me that draft.of tiddlers should probably be ignored anyway.


Also when installing from CPL, it installs tiddlyweb-sse as a dependancy, but the tiddlyweb-sse json is empty, so I think the install is broken. Is tiddlyweb-sse essential, or does it just handle updating the browser? Currently, I have modified `$:/config/SyncPollingInterval` to a one second sync interval for this.

TowelSniffer committed:
Fixes for externally created and deleted files.

- chokidar 4.0.3 uses modern Node.js imports which break the TiddlyWikii build tool so had to downgrade

!! Some modifications to fix issues with the sync logic

1. Fix duplicate Files (_1.tid) on external creation
2. Fix deletions not updating in browser
3. Fix crash on "Unknown File" deletion
4. Fix lockedFiles preventing re-deletion of same title
5. Fix "zombie" drafts

TowelSniffer committed:
Update FileSystemMonitor.ts

TowelSniffer committed:
Delete package-lock.json

TowelSniffer committed:
Update FileSystemMonitor.ts

linonetwo commented:
Thanks, while I'm not using this any more, as watch-fs feature is integrated into TidGi as a new filesystem sync adaptor. https://github.com/tiddly-gittly/TidGi-Desktop/tree/master/src/services/wiki/plugin/watchFileSystemAdaptor

And I have to say, maintaining this plugin may be a wrong direction. We should combine watch-fs with filesystem sync adaptor, otherwise there will be information gap, make filtering fs event difficult.

TowelSniffer committed:
Add test for changes.

- Add unit tests for race condition and crash fixes

- added 'Draft of' and dot files to regex ignore

TowelSniffer committed:
fixed sabotage

linonetwo committed:
Create pr-validate.yml

linonetwo committed:
upgrade deps

TowelSniffer committed:
Update test-watch-fs.js

TowelSniffer committed:
Merge branch 'master' of https://github.com/TowelSniffer/watch-fs

TowelSniffer committed:
Update test-watch-fs.js

added simulation deletion test

linonetwo commented:
I've released a new version, feel free to push fix if you encounter some.
